core/src/main/java/com/datastrato/gravitino/lock/TreeLockNode.java

```diff
@@ -0,0 +1,170 @@
+  public static final Logger LOG = LoggerFactory.getLogger(TreeLockNode.class);
+  private final String name;
+  private final ReentrantReadWriteLock readWriteLock;
+  @VisibleForTesting final Map<String, TreeLockNode> childMap;
+  private final LockManager lockManager;
+
+  // The reference count of this node. The reference count is used to track the number of the
+  // TreeLocks that are using this node. If the reference count is 0, it means that no TreeLock is
+  // using this node, and this node can be removed from the tree.
+  private final AtomicLong referenceCount = new AtomicLong();
+
+  protected TreeLockNode(String name, LockManager manager) {
+    this.name = name;
+    this.readWriteLock = new ReentrantReadWriteLock();
+    this.childMap = new ConcurrentHashMap<>();
+    this.lockManager = manager;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  /**
+   * Increase the reference count of this node. The reference count should always be greater than or
+   * equal to 0.
+   */
+  synchronized void addReference() {
+    referenceCount.getAndIncrement();
+  }
+
+  /**
+   * Decrease the reference count of this node. The reference count should always be greater than or
+   * equal to 0.
+   */
+  synchronized void decReference() {
+    referenceCount.getAndDecrement();
+  }
+
+  long getReference() {
+    return referenceCount.get();
+  }
+
+  /**
+   * Lock the node with the given lock type. This method should be followed by {@link
+   * #unlock(LockType)}.
+   *
+   * @param lockType The lock type to lock the node.
+   */
+  void lock(LockType lockType) {
+    if (lockType == LockType.READ) {
+      readWriteLock.readLock().lock();
+    } else {
+      readWriteLock.writeLock().lock();
+    }
+
+    LOG.trace("Node {} has been lock with '{}' lock", name, lockType);
+  }
+
+  /**
+   * Unlock the node with the given lock type. This method should be called after {@link
+   * #lock(LockType)}, and the lock type should be the same as the lock type in {@link
+   * #lock(LockType)}.
+   *
+   * @param lockType The lock type to unlock the node.
+   */
+  void unlock(LockType lockType) {
+    if (lockType == LockType.READ) {
+      readWriteLock.readLock().unlock();
+    } else {
+      readWriteLock.writeLock().unlock();
+    }
+
+    LOG.trace("Node {} has been unlock with '{}' lock", name, lockType);
+  }
+
+  /**
+   * Get the tree lock node by the given name. If the node doesn't exist, create a new TreeNode.
+   *
+   * <p>Note: This method should always be guarded by object lock.
+   *
+   * @param name The name of a resource such as entity or others.
+   * @return The tree lock node of this ident.
+   */
+  TreeLockNode getOrCreateChild(String name) {
+    TreeLockNode childNode =
+        childMap.computeIfAbsent(
+            name,
+            k -> {
+              // If the total node count is greater than the max node count, in case of memory
+              // leak and explosion, we should throw an exception.
```

Comment:
Introducing LockManager into TreeLockNode is inappropriate. It is recommended to handle it within LockManager. A pair can be returned, <node, isCreated>.

Reply:
OK, fixed